pkg/ocm/clustertransfer/cluster_transfer.go

```diff
@@ -0,0 +1,276 @@
+
+	var statusMsg string
+	updating, err := c.isUpdateRequired(data)
+	if err != nil {
+		statusMsg = fmt.Sprintf("pull-secret check failed: %v", err)
+		klog.Errorf(statusMsg)
+		c.updateStatus(false, statusMsg, "DataCorrupted", nil)
+		return
+	}
+	if updating {
+		err = c.updatePullSecret(data)
+		if err != nil {
+			statusMsg = fmt.Sprintf("Failed to update pull-secret: %v", err)
+			klog.Errorf(statusMsg)
+			c.updateStatus(false, statusMsg, "UpdateFailed", nil)
+			return
+		}
+		statusMsg = "pull-secret successfully updated"
+		klog.Info(statusMsg)
+	} else {
+		statusMsg = "no new data received"
+		klog.Info(statusMsg)
+	}
+	c.updateStatus(true, statusMsg, "", nil)
+}
+
+// isUpdateRequired checks if an update of the pull-secret is required or not.
+func (c *Controller) isUpdateRequired(newData []byte) (bool, error) {
+	pullSecret, err := c.getPullSecret()
+	if err != nil {
+		return false, err
+	}
+	c.pullSecret = pullSecret
+
+	newPullSecret, err := unmarshallToPullSecretContent(newData)
+	if err != nil {
+		return false, err
+	}
+	originalPullSecret, err := unmarshallToPullSecretContent(pullSecret.Data[v1.DockerConfigJsonKey])
+	if err != nil {
+		return false, err
+	}
+	return !isUpdatedPullSecretContentSame(*originalPullSecret, *newPullSecret), nil
+}
+
+// updatePullSecret creates a JSON merge patch of existing and new pull-secret data.
+// The result of the patch is used for pull-secret data update.
+func (c *Controller) updatePullSecret(newData []byte) error {
+	if c.pullSecret == nil {
+		ps, err := c.getPullSecret()
+		if err != nil {
+			return err
+		}
+		c.pullSecret = ps
+	}
+	existingData := c.pullSecret.Data[v1.DockerConfigJsonKey]
+
+	updatedData, err := jsonpatch.MergePatch(existingData, newData)
+	if err != nil {
+		return err
+	}
+
+	c.pullSecret.Data[v1.DockerConfigJsonKey] = updatedData
+	_, err = c.coreClient.Secrets("openshift-config").Update(c.ctx, c.pullSecret, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// requestClusterTransferWithExponentialBackoff queries a cluster transfer object
+// from the OCM API with exponential backoff.
+// It returns HttpError (see insightsclient.go) in case of any HTTP error response from the OCM API.
+// The exponential backoff is applied only for HTTP errors >= 500.
+func (c *Controller) requestClusterTransferWithExponentialBackoff(endpoint string) ([]byte, error) {
+	bo := wait.Backoff{
+		Duration: c.configurator.Config().OCMConfig.ClusterTransferInterval / 48, // 30 min as the first waiting
+		Factor:   2,
+		Jitter:   0,
+		Steps:    status.OCMAPIFailureCountThreshold,
+		Cap:      c.configurator.Config().OCMConfig.ClusterTransferInterval,
+	}
+
+	var data []byte
+	err := wait.ExponentialBackoff(bo, func() (bool, error) {
+		resp, err := c.client.Get(endpoint)
+		if err != nil {
+			return true, err
+		}
+		defer resp.Body.Close()
+		// TODO this will be moved to the insightsclient
```

Comment:
when PoC gets approved, convert it to an issue.

Reply:
Yes thanks. My plan is not to merge it. I will keep it here once there is a new OCM API implemented.